### ISSUE TITLE
Memory leak with PyPy 4.0.0

### ISSUE THREAD

**Issue by Armannas** [python-pillow/Pillow#1513]
I'm trying to use PyPy and Pillow in Docker. I modified the Dockerfile ( https://github.com/docker-library/pypy/blob/f1c1a5dd1248c044dafc52ddff7ff18be77d0ea8/2/Dockerfile ) to install PyPy 4.0.0 instead of 2.6.1 because the latter hangs at pip install Pillow indefinitely.

All works well, however when using code like:

``` python
for root, dirs, images in os.walk(IMAGE_DIR):
        for image in images:

            img_path = join(root, image)
            img = Image.open(img_path)
            img.save(img_path, optimize=False, quality=65)
```

The memory keeps building up and after completion does not get free'd.
I tried adding gc.collect() after every img.save() but that makes the code extremely slow, although helps keeping memory constant. Doing gc.collect() once after the entire os.walk loop is finished has no effect. When only opening the image followed by explicitly closing the image, the memory does not increase by a lot after rerunning the script many times.
Opening the image, followed by saving and ending with 'del img' and img.close() has no effect, memory is not free'd

Tried setting environment variables ENV PYPY_GC_MIN=1GB, ENV PYPY_GC_MAX=3GB in order to trigger the garbage collecting, did not work either.

It's not just Image.save, Image.rotate (and probably other methods) has the same effect.

**Comment by radarhere** [comment]
From the sound of it, I would guess that the problem lies in loading the image. Pillow lazy loads images, so calling open doesn't actually load the data, but save or rotate would.

You mentioned that opening and immediately closing does not cause the issue. I presume that just opening and leaving the image alone does not produce it either -

    for root, dirs, images in os.walk(IMAGE_DIR):
        for image in images:
            img_path = join(root, image)
            img = Image.open(img_path)

I presume that calling `open` and then `load` does.

    for root, dirs, images in os.walk(IMAGE_DIR):
        for image in images:
            img_path = join(root, image)
            img = Image.open(img_path)
            img.load()

If so, then that at least narrows down the problem a bit.

**Comment by Armannas** [comment]
> You mentioned that opening and immediately closing does not cause the issue. 
> I presume that just  opening and leaving the image alone does not produce it either 

I'll have to correct and elaborate myself on that one. 

I'm testing on ~4500 images and pressing F5 ten times or so, to rerun the script and look at my memory usage. My machine has 8gb.

- If I close the image immediately, the memory leak is very small. Within a percent of my memory, but still builds up and does not get free'd if I rerun my script a loooot of times.

- If I just open the image the memory leak is somewhat larger, a few (5-10)%

- your second scenario actually froze my machine. Running only once skyrockets to 97% RAM

- The same issue when using save,rotate etc. 

Had no trouble with CPython with his scenario.

**Comment by mattip** [comment]
While cpython uses a refcount on each object to manage memory allocation/deallocation, pypy uses a garbage collector. This difference is detailed in the compatibility section of the pypy documentation, an overview is available [here](http://pypy.org/compat.html). 

Could you write the code with Image() as a context manager::

    for root, dirs, images in os.walk(IMAGE_DIR):
            for image in images:
                img_path = join(root, image)
                with Image.open(img_path) as img:
                    img.save(img_path, optimize=False, quality=65)

**Comment by Armannas** [comment]
@mattip This solution really made sense, I read that piece of doc a million times and it never occured to me it is the same situation I am in. Unfortunately, this had no effect on my memory leakage. It is still rising even as a context manager and never frees the memory, unless I change my code or quit my Django server.

Very very very frustrating.

**Comment by arigo** [comment]
Looking at the source code of PIL, close() or __exit__() only forgets the underlying image object from the ``_imaging`` C extension.  It doesn't free it now.  Maybe it should be fixed so that it does.  In the meantime, you have no other solution than calling ``gc.collect()`` from time to time --- apart maybe the following, which is basically the same thing but only when internal counters are reached: ``import __pypy__; __pypy__.add_memory_pressure(n)`` where ``n`` is an estimate of the memory usage in bytes.

**Comment by wiredfool** [comment]
My understanding here is that this isn't technically a memory leak, it's just that the garbage collection is slow to get to destroying the original object when it's no longer reachable. In this case, the difference in implementation between the pypys is the issue, not the difference between reachability and refcounting as in PyPy vs CPython. (which tends to favor pypy, as it will free objects that we've forgotten to decref in the C layer where as they really leak in CPython).

As the simplest solution, does adding a `del` to the close method do it?:

```
    def close(self):
        """
        Closes the file pointer, if possible.

        This operation will destroy the image core and release its memory.
        The image data will be unusable afterward.

        This function is only required to close images that have not
        had their file read and closed by the
        :py:meth:`~PIL.Image.Image.load` method.
        """
        try:
            self.fp.close()
        except Exception as msg:
            logger.debug("Error closing: %s" % msg)

        del self.im
        # Instead of simply setting to None, we're setting up a
        # deferred error that will better explain that the core image
        # object is gone.
        self.im = deferred_error(ValueError("Operation on closed image"))
```

(that's in this branch, if you'd like an already patched version: https://github.com/wiredfool/Pillow/tree/pypy4_mem_leak)

Otherwise, we could add something in the `close`/`__exit__` layer that would forcibly destroy core of the `_imaging` object prior to the `Image` object destruction. It would essentially be the `_dealloc` method, but probably not including the call to PyObject_Del for the actual `_imaging` object. We'd have to make sure that further access to the `_imaging` object was prevented, as there are many accesses to it that do not check for a valid pointer first.

**Comment by Armannas** [comment]
@wiredfool Thanks for your effort. I used your patched version but the extra del didn't help unfortunately :( 

Is anyone else able to reproduce the problem?

**Comment by wiredfool** [comment]
I've added some code to free the core image memory prior to object deletion in the `__exit__`\`close` handler. 

Take another look at that branch and let me know.

**Comment by wiredfool** [comment]
@Armannas Did the new memory deletion change anything?

**Comment by Armannas** [comment]
@wiredfool going to try this a.s.a.p.

Thanks

**Comment by Armannas** [comment]
Still no luck. Memory builds up and doesn't get free'd. I need to stop docker or change something trivial in a Django file to free it.

**Comment by wiredfool** [comment]
Damn.

**Comment by aclark4life** [comment]
Damn indeed. Are we ready to call this a bug, or alternatively maybe "won't fix"?

**Comment by arigo** [comment]
Can someone comment on whether my proposed fix helps?  Adding ``import __pypy__; __pypy__.add_memory_pressure(number_of_bytes)`` e.g. in the ``load()`` method of PIL.

**Comment by hugovk** [comment]
PyPy2.7 and PyPy3.5 v6.0 have been released.

https://morepypy.blogspot.com/2018/04/pypy27-and-pypy35-v60-dual-release.html

Does this still occur with 6.0?

**Comment by hugovk** [comment]
PyPy2.7, PyPy3.5 and PyPy3.6-alpha v7.0 have been released.

https://morepypy.blogspot.com/2019/02/pypy-v700-triple-release-of-27-35-and.html

Does this still occur with 7.0?
